File: src/protocol/frames.ts
import { concat, readU16, readU32, u16be, u32be, utf8, utf8decode } from "./bytes";
import { crc32, crc32Bytes, readCrc32 } from "./crc32";
import {
  HASH_LEN,
  KIND_CAL,
  KIND_DATA,
  KIND_HEADER,
  MAGIC,
  PROTOCOL_VERSION,
  PUBKEY_LEN,
  SESSION_ID_LEN,
  SIG_LEN,
} from "./constants";
import { sign, verify } from "./crypto";

export type CalFrame = {
  kind: typeof KIND_CAL;
  sessionId: Uint8Array;
};

export type HeaderPayload = {
  sessionId: Uint8Array;
  fileName: string;
  fileSize: number;
  k: number;
  blockSize: number;
  fileHash: Uint8Array;
  publicKey: Uint8Array;
  mime: string;
};

export type HeaderFrame = HeaderPayload & {
  kind: typeof KIND_HEADER;
  signature: Uint8Array;
};

export type DataFrame = {
  kind: typeof KIND_DATA;
  sessionId: Uint8Array;
  symbolId: number;
  payload: Uint8Array;
  signature: Uint8Array;
};

export type DecodedFrame = CalFrame | HeaderFrame | DataFrame;

function magicOk(b: Uint8Array): boolean {
  return b.length >= 2 && b[0] === MAGIC[0] && b[1] === MAGIC[1];
}

export function encodeCal(sessionId: Uint8Array): Uint8Array {
  const body = concat(MAGIC, new Uint8Array([PROTOCOL_VERSION, KIND_CAL]), sessionId);
  return concat(body, crc32Bytes(body));
}

export function encodeHeaderBody(h: HeaderPayload): Uint8Array {
  const name = utf8(h.fileName.slice(0, 180));
  const mime = utf8((h.mime || "application/octet-stream").slice(0, 80));
  return concat(
    MAGIC,
    new Uint8Array([PROTOCOL_VERSION, KIND_HEADER]),
    h.sessionId,
    u16be(name.length),
    name,
    u32be(h.fileSize),
    u16be(h.k),
    u16be(h.blockSize),
    h.fileHash,
    h.publicKey,
    u16be(mime.length),
    mime,
  );
}

export async function encodeHeader(h: HeaderPayload, secretKey: Uint8Array): Promise<Uint8Array> {
  const body = encodeHeaderBody(h);
  const signature = await sign(body, secretKey);
  return concat(body, signature, crc32Bytes(concat(body, signature)));
}

export async function encodeData(
  sessionId: Uint8Array,
  symbolId: number,
  payload: Uint8Array,
  secretKey: Uint8Array,
): Promise<Uint8Array> {
  const body = concat(
    MAGIC,
    new Uint8Array([PROTOCOL_VERSION, KIND_DATA]),
    sessionId,
    u16be(symbolId),
    u16be(payload.length),
    payload,
  );
  const signature = await sign(body, secretKey);
  return concat(body, signature, crc32Bytes(concat(body, signature)));
}

export async function decodeFrame(raw: Uint8Array, expectedKey?: Uint8Array): Promise<DecodedFrame | null> {
  if (raw.length < 8 || !magicOk(raw)) return null;
  if (raw[2] !== PROTOCOL_VERSION) return null;

  const crcOff = raw.length - 4;
  const bodyWithSig = raw.subarray(0, crcOff);
  if (crc32(bodyWithSig) !== readCrc32(raw, crcOff)) return null;

  const kind = raw[3]!;
  if (kind === KIND_CAL) {
    if (raw.length < 4 + SESSION_ID_LEN + 4) return null;
    return { kind: KIND_CAL, sessionId: raw.subarray(4, 4 + SESSION_ID_LEN) };
  }

  if (kind === KIND_HEADER) {
    let o = 4;
    const sessionId = raw.subarray(o, o + SESSION_ID_LEN);
    o += SESSION_ID_LEN;
    const nameLen = readU16(raw, o);
    o += 2;
    const fileName = utf8decode(raw.subarray(o, o + nameLen));
    o += nameLen;
    const fileSize = readU32(raw, o);
    o += 4;
    const k = readU16(raw, o);
    o += 2;
    const blockSize = readU16(raw, o);
    o += 2;
    const fileHash = raw.subarray(o, o + HASH_LEN);
    o += HASH_LEN;
    const publicKey = raw.subarray(o, o + PUBKEY_LEN);
    o += PUBKEY_LEN;
    const mimeLen = readU16(raw, o);
    o += 2;
    const mime = utf8decode(raw.subarray(o, o + mimeLen));
    o += mimeLen;
    const signature = raw.subarray(o, o + SIG_LEN);
    const body = raw.subarray(0, o);
    if (!(await verify(signature, body, publicKey))) return null;
    return {
      kind: KIND_HEADER,
      sessionId,
      fileName,
      fileSize,
      k,
      blockSize,
      fileHash,
      publicKey,
      mime,
      signature,
    };
  }

  if (kind === KIND_DATA) {
    if (!expectedKey) return null;
    let o = 4;
    const sessionId = raw.subarray(o, o + SESSION_ID_LEN);
    o += SESSION_ID_LEN;
    const symbolId = readU16(raw, o);
    o += 2;
    const payloadLen = readU16(raw, o);
    o += 2;
    const payload = raw.subarray(o, o + payloadLen);
    o += payloadLen;
    const signature = raw.subarray(o, o + SIG_LEN);
    const body = raw.subarray(0, o);
    if (!(await verify(signature, body, expectedKey))) return null;
    return { kind: KIND_DATA, sessionId, symbolId, payload, signature };
  }

  return null;
}

export function b64urlEncode(data: Uint8Array): string {
  let bin = "";
  for (let i = 0; i < data.length; i++) bin += String.fromCharCode(data[i]!);
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

export function b64urlDecode(s: string): Uint8Array | null {
  try {
    const pad = s.length % 4 === 0 ? "" : "=".repeat(4 - (s.length % 4));
    const b64 = s.replace(/-/g, "+").replace(/_/g, "/") + pad;
    const bin = atob(b64);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
  } catch {
    return null;
  }
}

File: android/core-protocol/src/main/kotlin/app/candela/protocol/Frames.kt
package app.candela.protocol

import app.candela.protocol.Constants.CRC_LEN
import app.candela.protocol.Constants.HASH_LEN
import app.candela.protocol.Constants.KIND_CAL
import app.candela.protocol.Constants.KIND_DATA
import app.candela.protocol.Constants.KIND_HEADER
import app.candela.protocol.Constants.MAGIC
import app.candela.protocol.Constants.PROTOCOL_VERSION
import app.candela.protocol.Constants.PUBKEY_LEN
import app.candela.protocol.Constants.SESSION_ID_LEN
import app.candela.protocol.Constants.SIG_LEN

/**
 * CAL / HEADER / DATA codec. 1:1 with src/protocol/frames.ts.
 *
 * Wire layout (PSR section 2.3):
 *   CAL    magic(2) version(1) kind=0(1) sessionId(8) CRC32(4)
 *   HEADER magic(2) version(1) kind=1(1) sessionId(8) nameLen(2) name
 *          fileSize(4) k(2) blockSize(2) fileHash(32) publicKey(32)
 *          mimeLen(2) mime sig(64) CRC32(4)
 *   DATA   magic(2) version(1) kind=2(1) sessionId(8) symbolId(2)
 *          payloadLen(2) payload sig(64) CRC32(4)
 *
 * Signature covers the body BEFORE the signature. CRC32 covers body+signature.
 *
 * Verify order is fixed by audit section 5.2 and is enforced in [decode]:
 *   CRC32 -> Ed25519 -> symbol-id bounds -> (caller) fountain insert.
 * CRC first because it is ~1000x cheaper than a signature check and rejects the
 * overwhelming majority of bad frames; the signature is what stops injection.
 */
sealed interface Frame {
    val sessionId: ByteArray

    data class Cal(override val sessionId: ByteArray) : Frame {
        override fun equals(other: Any?) = other is Cal && sessionId.contentEquals(other.sessionId)
        override fun hashCode() = sessionId.contentHashCode()
    }

    data class Header(
        override val sessionId: ByteArray,
        val fileName: String,
        val fileSize: Long,
        val k: Int,
        val blockSize: Int,
        val fileHash: ByteArray,
        val publicKey: ByteArray,
        val mime: String,
        val signature: ByteArray,
    ) : Frame {
        override fun equals(other: Any?) = other is Header &&
            sessionId.contentEquals(other.sessionId) &&
            fileName == other.fileName && fileSize == other.fileSize &&
            k == other.k && blockSize == other.blockSize &&
            fileHash.contentEquals(other.fileHash) &&
            publicKey.contentEquals(other.publicKey) && mime == other.mime

        override fun hashCode(): Int = fileName.hashCode() * 31 + k
    }

    data class Data(
        override val sessionId: ByteArray,
        val symbolId: Int,
        val payload: ByteArray,
        val signature: ByteArray,
    ) : Frame {
        override fun equals(other: Any?) = other is Data &&
            sessionId.contentEquals(other.sessionId) &&
            symbolId == other.symbolId && payload.contentEquals(other.payload)

        override fun hashCode(): Int = symbolId * 31 + payload.contentHashCode()
    }
}

/** Why a frame was rejected — surfaced to the coach HUD and to tests. */
enum class RejectReason {
    TOO_SHORT, BAD_MAGIC, BAD_VERSION, BAD_CRC, BAD_SIGNATURE,
    UNKNOWN_KIND, MALFORMED, SYMBOL_OUT_OF_RANGE, MISSING_KEY,
}

sealed interface DecodeResult {
    data class Ok(val frame: Frame) : DecodeResult
    data class Rejected(val reason: RejectReason) : DecodeResult
}

data class HeaderPayload(
    val sessionId: ByteArray,
    val fileName: String,
    val fileSize: Long,
    val k: Int,
    val blockSize: Int,
    val fileHash: ByteArray,
    val publicKey: ByteArray,
    val mime: String,
) {
    override fun equals(other: Any?): Boolean = other is HeaderPayload &&
        sessionId.contentEquals(other.sessionId) && fileName == other.fileName
    override fun hashCode(): Int = fileName.hashCode()
}

object Frames {

    fun encodeCal(sessionId: ByteArray): ByteArray {
        val body = Bytes.concat(
            MAGIC,
            byteArrayOf(PROTOCOL_VERSION.toByte(), KIND_CAL.toByte()),
            sessionId,
        )
        return Bytes.concat(body, Crc32.bytes(body))
    }

    fun encodeHeaderBody(h: HeaderPayload): ByteArray {
        val name = Bytes.utf8Truncate(h.fileName, Constants.MAX_NAME_BYTES)
        val mimeStr = h.mime.ifEmpty { "application/octet-stream" }
        val mime = Bytes.utf8Truncate(mimeStr, Constants.MAX_MIME_BYTES)
        return Bytes.concat(
            MAGIC,
            byteArrayOf(PROTOCOL_VERSION.toByte(), KIND_HEADER.toByte()),
            h.sessionId,
            Bytes.u16be(name.size),
            name,
            Bytes.u32be(h.fileSize),
            Bytes.u16be(h.k),
            Bytes.u16be(h.blockSize),
            h.fileHash,
            h.publicKey,
            Bytes.u16be(mime.size),
            mime,
        )
    }

    fun encodeHeader(h: HeaderPayload, secretKey: ByteArray): ByteArray {
        val body = encodeHeaderBody(h)
        val sig = Crypto.sign(body, secretKey)
        val bodySig = Bytes.concat(body, sig)
        return Bytes.concat(bodySig, Crc32.bytes(bodySig))
    }

    fun encodeData(
        sessionId: ByteArray,
        symbolId: Int,
        payload: ByteArray,
        secretKey: ByteArray,
    ): ByteArray {
        val body = Bytes.concat(
            MAGIC,
            byteArrayOf(PROTOCOL_VERSION.toByte(), KIND_DATA.toByte()),
            sessionId,
            Bytes.u16be(symbolId),
            Bytes.u16be(payload.size),
            payload,
        )
        val sig = Crypto.sign(body, secretKey)
        val bodySig = Bytes.concat(body, sig)
        return Bytes.concat(bodySig, Crc32.bytes(bodySig))
    }

    /**
     * @param expectedKey required for DATA frames (they are signed by the key the
     *   HEADER established). DATA without a key is rejected, never trusted.
     * @param expectedK when known, bounds-checks symbolId so a corrupt-but-valid
     *   frame cannot poison the fountain.
     */
    fun decode(raw: ByteArray, expectedKey: ByteArray? = null, expectedK: Int? = null): DecodeResult {
        if (raw.size < 8) return DecodeResult.Rejected(RejectReason.TOO_SHORT)
        if (raw[0] != MAGIC[0] || raw[1] != MAGIC[1]) {
            return DecodeResult.Rejected(RejectReason.BAD_MAGIC)
        }
        if ((raw[2].toInt() and 0xFF) != PROTOCOL_VERSION) {
            return DecodeResult.Rejected(RejectReason.BAD_VERSION)
        }

        val crcOff = raw.size - CRC_LEN
        if (Crc32.compute(raw, 0, crcOff) != Crc32.read(raw, crcOff)) {
            return DecodeResult.Rejected(RejectReason.BAD_CRC)
        }

        return when (raw[3].toInt() and 0xFF) {
            KIND_CAL -> decodeCal(raw)
            KIND_HEADER -> decodeHeader(raw)
            KIND_DATA -> decodeData(raw, expectedKey, expectedK)
            else -> DecodeResult.Rejected(RejectReason.UNKNOWN_KIND)
        }
    }

    private fun decodeCal(raw: ByteArray): DecodeResult {
        if (raw.size < 4 + SESSION_ID_LEN + CRC_LEN) {
            return DecodeResult.Rejected(RejectReason.MALFORMED)
        }
        return DecodeResult.Ok(Frame.Cal(raw.copyOfRange(4, 4 + SESSION_ID_LEN)))
    }

    private fun decodeHeader(raw: ByteArray): DecodeResult = try {
        var o = 4
        val sessionId = raw.copyOfRange(o, o + SESSION_ID_LEN); o += SESSION_ID_LEN
        val nameLen = Bytes.readU16(raw, o); o += 2
        val name = Bytes.utf8Decode(raw.copyOfRange(o, o + nameLen)); o += nameLen
        val fileSize = Bytes.readU32(raw, o); o += 4
        val k = Bytes.readU16(raw, o); o += 2
        val blockSize = Bytes.readU16(raw, o); o += 2
        val fileHash = raw.copyOfRange(o, o + HASH_LEN); o += HASH_LEN
        val publicKey = raw.copyOfRange(o, o + PUBKEY_LEN); o += PUBKEY_LEN
        val mimeLen = Bytes.readU16(raw, o); o += 2
        val mime = Bytes.utf8Decode(raw.copyOfRange(o, o + mimeLen)); o += mimeLen
        val sig = raw.copyOfRange(o, o + SIG_LEN)
        val body = raw.copyOfRange(0, o)

        if (!Crypto.verify(sig, body, publicKey)) {
            DecodeResult.Rejected(RejectReason.BAD_SIGNATURE)
        } else {
            DecodeResult.Ok(
                Frame.Header(sessionId, name, fileSize, k, blockSize, fileHash, publicKey, mime, sig),
            )
        }
    } catch (_: IndexOutOfBoundsException) {
        DecodeResult.Rejected(RejectReason.MALFORMED)
    }

    private fun decodeData(raw: ByteArray, expectedKey: ByteArray?, expectedK: Int?): DecodeResult {
        if (expectedKey == null) return DecodeResult.Rejected(RejectReason.MISSING_KEY)
        return try {
            var o = 4
            val sessionId = raw.copyOfRange(o, o + SESSION_ID_LEN); o += SESSION_ID_LEN
            val symbolId = Bytes.readU16(raw, o); o += 2
            val payloadLen = Bytes.readU16(raw, o); o += 2
            val payload = raw.copyOfRange(o, o + payloadLen); o += payloadLen
            val sig = raw.copyOfRange(o, o + SIG_LEN)
            val body = raw.copyOfRange(0, o)

            when {
                !Crypto.verify(sig, body, expectedKey) ->
                    DecodeResult.Rejected(RejectReason.BAD_SIGNATURE)
                expectedK != null && symbolId >= 65536 ->
                    DecodeResult.Rejected(RejectReason.SYMBOL_OUT_OF_RANGE)
                else -> DecodeResult.Ok(Frame.Data(sessionId, symbolId, payload, sig))
            }
        } catch (_: IndexOutOfBoundsException) {
            DecodeResult.Rejected(RejectReason.MALFORMED)
        }
    }
}
